### PR TITLE
feat(v2): update swizzle command to suggest component/theme

### PR DESCRIPTION
**Pull request by anshulrgoyal**
## Motivation
Better user experience for swizzle cli. Based on #2984

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan
image
image
image
image
image

### CODE REVIEW

**Comment by slorber** [comment]
That looks like a nice improvement.

What about making the message clearer like "this is the list of available themes you can swizzle" or something?

**Comment by anshulrgoyal** [comment]
> That looks like a nice improvement.
> 
> What about making the message clearer like "this is the list of available themes you can swizzle" or something?

image 

Something like this?

**Comment by slorber** [comment]
yes, that looks fine thanks

**Comment by slorber** [comment]
Thanks @anshulrgoyal that looks like a first version.

Here are a few suggestions to improve:
- When component/theme not found, and no suggestion, show again the list of available components
- Show components to swizzle in subfolders too (for example, it's a bit weird to suggest `hooks`, user is more likely to swizzle a single hook instead of the whole folder, and the following is actually working: `yarn swizzle @docusaurus/theme-classic hooks/useAnnouncementBar`)

**Comment by slorber** [comment]
I wonder if also the theme should not explicitly define the component list to swizzle.

Because currently, some theme components are considered "implementation details", and we don't necessarily want to encourage the users to swizzle those, as this would implicitly expand the api surface of a theme.

What about exploring the idea of defining, at the theme level, a list of swizzle suggestions? 

We could allow the user to swizzle values out of the suggestions, with a flag like --danger or something, so that they know that they are swizzling internal comp, and they may have to handle breaking changes

**Comment by anshulrgoyal** [comment]
> Thanks @anshulrgoyal that looks like a first version.
> 
> Here are a few suggestions to improve:
> 
>     * When component/theme not found, and no suggestion, show again the list of available components
> 
>     * Show components to swizzle in subfolders too (for example, it's a bit weird to suggest `hooks`, user is more likely to swizzle a single hook instead of the whole folder, and the following is actually working: `yarn swizzle @docusaurus/theme-classic hooks/useAnnouncementBar`)

I will add a tree-like listing for components and idea of theme defining swizzle component is nice but It would be a breaking change :), we can let user export a list of swizzle component.

**Comment by anshulrgoyal** [comment]
Or we can let the theme define a list of dangerous component

**Comment by anshulrgoyal** [comment]
Something like this works @slorber?
image

**Comment by anshulrgoyal** [comment]
For component not found:
image
If no suggestion:
image

**Comment by slorber** [comment]
thanks that looks better

> I will add a tree-like listing for components and idea of theme defining swizzle component is nice but It would be a breaking change :), we can let user export a list of swizzle component.

Not sure what you mean by breaking changes? it's just a list of suggestions displayed, it should not forbid the user to swizzle a component that is internal, just not suggest to swizzle it. This feature should not affect current swizzle behavior, it's just a list of suggestions


> Or we can let the theme define a list of dangerous component

I'd rather have the opposite instead. By default comps are considered as internal code, unless we explicitly add them to the list, is less risky than having them being recommended for swizzling by default. Devs will likely forget to add comps to that list. If this happens, it's not a big deal, the comp won't be suggested in the list but the swizzle command on this comp will still work anyway.

**Comment by slorber** [comment]
Maybe we should add a method like isSwizzleRecommendation(compName: string): boolean or something, and let the theme decide if it wants to do a allow/denylist or whatever?

I think the classic theme should rather be an allow list to avoid recommending to swizzle impl details, but for other smaller themes we might as well want to suggest all comps by default, and avoid annoying the user to manually craft a list of allowed comps to swizzle

**Comment by anshulrgoyal** [comment]
> thanks that looks better
> 
> > I will add a tree-like listing for components and idea of theme defining swizzle component is nice but It would be a breaking change :), we can let user export a list of swizzle component.
> 
> Not sure what you mean by breaking changes? it's just a list of suggestions displayed, it should not forbid the user to swizzle a component that is internal, just not suggest to swizzle it. This feature should not affect current swizzle behavior. Anyway even if we did forbid the user to swizzle with the cli, he would still be able to copy the files manually if he really wants to, and that would work.
> 
> > Or we can let the theme define a list of dangerous component
> 
> I'd rather have the opposite instead. By default comps are considered as internal code, unless we explicitly add them to the list, is less risky than having them being recommended for swizzling by default. Devs will likely forget to add comps to that list. If this happens, it's not a big deal, the comp won't be suggested in the list but the swizzle command on this comp will still work anyway.

I thought u meant to not allow swizzle of certain component in a theme.

**Comment by anshulrgoyal** [comment]
> Maybe we should add a method like isSwizzleRecommendation(compName: string): boolean or something, and let the theme decide if it wants to do a allow/denylist or whatever?
> 
> I think the classic theme should rather be an allow list to avoid recommending to swizzle impl details, but for other smaller themes we might as well want to suggest all comps by default, and avoid annoying the user to manually craft a list of allowed comps to swizzle

I am more inclined to list of allowed component because it will give more control to `@docusaurus/core` which can handle it in a more  uniform method. If we go with `isSwizzleRecommendation` we have to call it with every component to generate a list.

**Comment by anshulrgoyal** [comment]
image
image
image

@slorber  something like this?

**Comment by anshulrgoyal** [comment]
```js
const components=[  // 'AnnouncementBar',
  // 'BlogListPage',
  // 'BlogListPaginator',
  // 'BlogPostItem',
  // 'BlogPostPage',
  // 'BlogPostPaginator',
  // 'BlogTagsListPage',
  // 'BlogTagsPostsPage',
  'CodeBlock',
  // 'DocItem',
  // 'DocPage',
  // 'DocPaginator',
  'DocSidebar',
  'Footer',
  // 'Heading',
  // 'Layout',
  // 'MDXComponents',
  // 'Navbar',
  'NotFound',
  'SearchBar',
  // 'TabItem',
  // 'Tabs',
  // 'ThemeContext',
  // 'ThemeProvider',
  // 'Toggle',
  // 'UserPreferencesContext',
  // 'UserPreferencesProvider',
  // 'hooks/useAnnouncementBar',
  // 'hooks/useHideableNavbar',
  // 'hooks/useLocationHash',
  // 'hooks/useLockBodyScroll',
  // 'hooks/useLogo',
  // 'hooks/usePrismTheme',
  // 'hooks/useScrollPosition',
  // 'hooks/useTOCHighlight',
  // 'hooks/useTabGroupChoice',
  'hooks/useTheme',
  // 'hooks/useThemeContext',
  // 'hooks/useUserPreferencesContext',
  // 'hooks/useWindowSize',
  'prism-include-languages',]

```

File: website/src/theme/hooks/useTheme.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+/**
```
Comment (slorber) [inline website/src/theme/hooks/useTheme.js:1]: this should probably not be commited

**Comment by teikjun** [comment]
@anshulrgoyal btw, there's a merge conflict here

**Comment by anshulrgoyal** [comment]
> @anshulrgoyal btw, there's a merge conflict here

Thanks. I will resolve them

File: packages/docusaurus/src/commands/swizzle.ts [hunk 2]
```diff
@@ -9,61 +9,263 @@ import chalk = require('chalk');
...
 import importFresh from 'import-fresh';
 import path from 'path';
-import {Plugin, LoadContext} from '@docusaurus/types';
+import {Plugin, LoadContext, PluginConfig} from '@docusaurus/types';
+import leven from 'leven';
 
 import {THEME_PATH} from '../constants';
-import {loadContext} from '../server';
+import {loadContext, loadPluginConfigs} from '../server';
+import initPlugins, {pluginOptionsValidator} from '../server/plugins/init';
```
Comment (slorber) [inline packages/docusaurus/src/commands/swizzle.ts:17]:
```suggestion
import initPlugins from '../server/plugins/init';
import {normalizePluginOptions} from '@docusaurus/utils-validation';
```

File: packages/docusaurus/src/commands/swizzle.ts [hunk 3]
```diff
@@ -9,61 +9,264 @@ import chalk = require('chalk');
...
+            const [, options] = pluginConfig;
+            pluginOptions = options;
+          }
+        }
+      }
+    });
+    if (validateOptions) {
+      // normilize options
+      const normalizedOptions = validateOptions({
+        validate: pluginOptionsValidator,
```
Comment (slorber) [inline packages/docusaurus/src/commands/swizzle.ts:190]:
```suggestion
        validate: normalizePluginOptions,
```

**Comment by slorber** [comment]
that looks good thanks

**Pull request merged**